### ISSUE TITLE
SourceKit LSP (Swift) stops working shortly after regular use

### ISSUE THREAD

**Issue by GhIlyusha** [helix-editor/helix#6334]
### Summary

After a short amount of time editing a basic Swift file LSP features like autocompletion and diagnostics stop working and an error that reads: `Async job failed: request timed out` starts to frequently appear below the mode line.

### Reproduction Steps

Attempt to edit any swift file, try to use autocompletion several times (five at most).
It works for the first couple attempts, then it stops working.
I use SourceKit bundled with Xcode 14.2 (14C18).
Can't reproduce this with VSCode using the official Swift plugin.

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2023-03-16T18:48:12.924 helix_lsp::transport [ERROR] err: <- StreamClosed
2023-03-16T18:48:12.924 helix_term::application [DEBUG] received editor event: LanguageServerMessage((0, Notification(Notification { jsonrpc: None, method: "exit", params: None })))
2023-03-16T18:48:12.924 helix_view::editor [DEBUG] editor status: Language server exited
2023-03-16T18:48:12.924 helix_view::document [DEBUG] id 2 modified - last saved: 0, current: 0
2023-03-16T18:48:13.197 helix_term::application [DEBUG] received editor event: IdleTimer
2023-03-16T18:48:13.198 helix_view::document [DEBUG] id 2 modified - last saved: 0, current: 0
2023-03-16T18:48:13.462 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":40,"line":42},"start":{"character":40,"line":42}},"text":"D"}],"textDocument":{"uri":"file:///FILE.swift","version":130}}}
2023-03-16T18:48:13.463 helix_lsp::transport [ERROR] err: <- IO(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
2023-03-16T18:48:13.475 helix_vcs::diff::worker [INFO] Diff computation timed out, update of diffs might appear delayed
2023-03-16T18:48:13.477 helix_view::document [DEBUG] id 2 modified - last saved: 0, current: 0
2023-03-16T18:48:13.647 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":41,"line":42},"start":{"character":41,"line":42}},"text":"o"}],"textDocument":{"uri":"file:///FILE.swift","version":131}}}
2023-03-16T18:48:13.647 helix_lsp::transport [ERROR] err: <- IO(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
2023-03-16T18:48:13.658 helix_vcs::diff::worker [INFO] Diff computation timed out, update of diffs might appear delayed
2023-03-16T18:48:13.660 helix_view::document [DEBUG] id 2 modified - last saved: 0, current: 0
2023-03-16T18:48:13.886 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":42,"line":42},"start":{"character":42,"line":42}},"text":"u"}],"textDocument":{"uri":"file:///FILE.swift","version":132}}}
2023-03-16T18:48:13.886 helix_lsp::transport [ERROR] err: <- IO(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
```

</details>

### Platform

macOS 13.2.1

### Terminal Emulator

iTerm2 3.4.19

### Helix Version

helix 22.12 (96ff64a8)

**Comment by GhIlyusha** [comment]
@pascalkuthe Thank you for the information. I compiled the latest helix commit (21a3fb8f) and can reproduce the issue in the exact same way.
The error message changed to `Async job failed: request <number> timed out ` where number starts 4 or 5 and increases by 1 every time.
This can be reproduced in any newly created package (made with `swift package init`), as well as an Xcode project or a standalone file.
<img width="676" alt="Screenshot 2023-03-16 at 8 10 36 PM" src="https://user-images.githubusercontent.com/44343874/225698672-970ecf91-eb9b-4932-bc83-b5482669a4c7.png">

**Comment by the-mikedavis** [comment]
Can you run with `hx -v --log log.txt` and attach the `log.txt` file after opening a Swift file? That will show the communication between Helix and the language server.

**Comment by GhIlyusha** [comment]
@the-mikedavis Here is the log:
[log.txt](https://github.com/helix-editor/helix/files/10995630/log.txt)

**Comment by the-mikedavis** [comment]
It looks like all is going well until the server silently exits on L219 of the log. If sourcekit has a logfile it would be good to consult that as well. From what I see in the log this looks like a sourcekit issue rather than a helix issue

**Comment by GhIlyusha** [comment]
Unfortunately sourcekit doesn't appear to provide logfiles. It might not be an issue on helix's side, but both the neovim and the vscode lsp integrations (listed [here](https://github.com/apple/sourcekit-lsp/tree/main/Editors)) don't have this behavior for me (they also show autocompletion way faster..?)

Could helix at least attempt to restart the language server in a situation like this, maybe with a backoff?

**Comment by the-mikedavis** [comment]
I will try to reproduce this. In the meantime, all language server stderr is printed to the Helix log as ERROR level messages so you should be able to see sourcekit's logging if you set that environment variable even without running in verbose mode (`hx` rather than `hx -v`). You could add that key to the environment config for swift:

```toml
# ~/.config/helix/languages.toml
[[language]]
name = "swift"
language-server.environment.SOURCEKIT_LOGGING = "2"
```

**Comment by the-mikedavis** [comment]
I couldn't reproduce this with sourcekit-lsp and swift 5.7.3 on Linux

> Could helix at least attempt to restart the language server in a situation like this, maybe with some backoff to avoid infinite loops?

This would be a nice improvement. Currently you can restart the server manually with the `:lsp-restart` command.

**Comment by GhIlyusha** [comment]
Okay I can now confirm this is a bug with SourceKit. The issue is quite easy to repro for me by using the following code:
```swift
let x = UnsafeRawPointer()
let y = Doub//le()
```
So the actual resolution that I would appreciate here is automatic restarting of language server. Here's how vscode handles a restart loop:
<img width="499" alt="Screenshot 2023-03-21 at 3 12 33 PM" src="https://user-images.githubusercontent.com/44343874/226602348-70d25c94-5aef-421d-aa16-861738cbcf6f.png">

**Comment by the-mikedavis** [comment]
I'm confused with the swift versioning so I don't know if I'm ahead or behind you in version but on 5.7.3, I still can't reproduce with that case: https://asciinema.org/a/568984.

The log here ([log.txt](https://github.com/helix-editor/helix/files/11033990/log.txt)) has an error on virtually ever request after the first completion. It looks like sourcekit is having a hard time figuring out a TextEdit to suggest. Maybe in other versions of sourcekit that failure is causing an exit?

I'll make a separate issue for restarting the language server automatically so that it doesn't get mixed up with the debugging here.

**Comment by the-mikedavis** [comment]
Oh I see, they say that was fixed in 5.7.2 but that isn't released yet on macos

**Comment by GhIlyusha** [comment]
I'm guessing this is superseded by https://github.com/helix-editor/helix/issues/6398 then.
